modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3PerProjectClientManager.java

```diff
@@ -0,0 +1,283 @@
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.repositories.s3;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.SubscribableListener;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateApplier;
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.cluster.metadata.ProjectMetadata;
+import org.elasticsearch.common.settings.ProjectSecrets;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class S3PerProjectClientManager implements ClusterStateApplier {
+
+    private static final Logger logger = LogManager.getLogger(S3PerProjectClientManager.class);
+    private static final String S3_SETTING_PREFIX = "s3.";
+
+    private final Settings nodeS3Settings;
+    private final Function<S3ClientSettings, AmazonS3Reference> clientBuilder;
+    private final Executor executor;
+    // A map of projectId to clients holder. Adding to and removing from the map happen only in the applier thread.
+    private final Map<ProjectId, ClientsHolder> projectClientsHolders;
+    // Listener for tracking ongoing async closing of obsolete clients. Updated only in the applier thread.
+    private volatile SubscribableListener<Void> clientsCloseListener = null;
+
+    S3PerProjectClientManager(Settings settings, Function<S3ClientSettings, AmazonS3Reference> clientBuilder, Executor executor) {
+        this.nodeS3Settings = Settings.builder()
+            .put(settings.getByPrefix(S3_SETTING_PREFIX), false) // not rely on any cluster scoped secrets
+            .normalizePrefix(S3_SETTING_PREFIX)
+            .build();
+        this.clientBuilder = clientBuilder;
+        this.executor = executor;
+        this.projectClientsHolders = new ConcurrentHashMap<>();
+    }
+
+    // visible for tests
+    Map<ProjectId, ClientsHolder> getProjectClientsHolders() {
+        return Map.copyOf(projectClientsHolders);
+    }
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
```

Comment:
Change this to be an applier instead of a listener so that the repository (also created in an applier) is fully functional in any listeners of the same cluster state. This is necessary so that the repository can be used to bootstrapping a new project.